### PR TITLE
[Pipelining] Reorder _Action from F1_1 to 1F1

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129810
* __->__ #129786

Also steers away from accesing _Action via positional unpacking since
that is error prone

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @chauhang @d4l3k

### CODE REVIEW

**Comment by awgu** [comment]
Just so we do not forget: We need to experiment with/evaluate whether it is okay or not that if we run `unshard(async_op=True)` today, the all-gather copy-in ops will run in the current/default stream, so they cannot overlap with compute. If we need to allow this, then we need some changes or some other API, but there is some nuance there (I can explain later).

**Comment by kwen2501** [comment]
Can you share more context on the need of adding UNSHARD/RESHARD actions? Is this a more advanced form of pipeline + FSDP so that people can have finer grain control or better overlap?

**Comment by awgu** [comment]
@kwen2501 yes :)

**Review by H-Huang — APPROVED** [review]

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -30,15 +30,20 @@
...
     WEIGHT = 3
+    UNSHARD = 4
+    RESHARD = 5
 
     def __str__(self):
         str_map = {
             _ComputationType.FORWARD: "F",
             _ComputationType.BACKWARD: "B",
             _ComputationType.WEIGHT: "W",
+            _ComputationType.UNSHARD: "UNSHARD",
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:47]: i almost did that but i decided against.
* F,B,W is pretty easy to follow and its nice to have single-letters for the simple layer of the IR
* but once we add a lot of things we'll start hitting conflicts.  I plan to add Send/Recv next, and R would be recv or reshard.  I think for the lower layer of IR it is better to not try to be so brief.

**Comment by wconstab** [comment]
FYI I split the original PR so this one is just focusing on the refactor of _Action class.  Then i add UNSHARD/RESHARD stuff in the next PR.

**Pull request closed without merging**